pkg/controller/direct/logging/logmetric_controller.go

```diff
@@ -0,0 +1,253 @@
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &logMetricModel{}
+
+type logMetricAdapter struct {
+	resourceID string
+	parentID   string
+
+	desired         *krm.LoggingLogMetric
+	actual          *api.LogMetric
+	logMetricClient *api.ProjectsMetricsService
+}
+
+var _ directbase.Adapter = &logMetricAdapter{}
+
+// AdapterForObject implements the Model interface.
+func (m *logMetricModel) AdapterForObject(ctx context.Context, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	projectMetricsService, err := m.newProjectMetricsService(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	obj := &krm.LoggingLogMetric{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	if obj.Spec.ProjectRef.External == "" {
+		// todo acpana: resolve the external ref
+		return nil, fmt.Errorf("project external reference is not set")
+	}
+
+	// validate that the project ref follows the "projects/PROJECT_ID" format
+	if parts := strings.Split(obj.Spec.ProjectRef.External, "/"); len(parts) != 2 || parts[0] != "projects" {
+		return nil, fmt.Errorf("project external reference is not in the format projects/PROJECT_ID")
+	}
+
+	return &logMetricAdapter{
+		resourceID:      ValueOf(obj.Spec.ResourceID),
+		parentID:        obj.Spec.ProjectRef.External,
+		desired:         obj,
+		logMetricClient: projectMetricsService,
+	}, nil
+}
+
+func (a *logMetricAdapter) Find(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	logMetric, err := a.logMetricClient.Get(a.fullyQualifiedName()).Context(ctx).Do()
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting logMetric %q: %w", a.fullyQualifiedName(), err)
+	}
+
+	a.actual = logMetric
+
+	return true, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *logMetricAdapter) Delete(ctx context.Context) (bool, error) {
+	// Already deleted
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	_, err := a.logMetricClient.Delete(a.fullyQualifiedName()).Context(ctx).Do()
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("deleting log metric %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	return true, nil
+}
+
+func (a *logMetricAdapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating object", "u", u)
+
+	project := a.desired.Spec.ProjectRef.External
+	// todo acpana this looks like a good candidate for factored out validation;
+	// a shared validator? validate exists/ check set? validate is well formed?
+	if project == "" {
+		return fmt.Errorf("project is empty")
+	}
+	name := a.desired.GetName()
+	if name == "" {
+		return fmt.Errorf("name is empty")
+	}
+	filter := a.desired.Spec.Filter
+	if filter == "" {
+		return fmt.Errorf("filter is empty")
```

Comment:
No objection to doing this here, though I suspect over time we want to refactor validation into its own method.  Letting the GCP API return an error is always an option also (although the error messages tend not to be as good as if we hand-crafted them, as long as we keep the field names reasonably close the error messages tend not to be _bad_)